vaadin-text-field-flow-parent/vaadin-text-field-flow-integration-tests/src/test/java/com/vaadin/flow/component/textfield/tests/validation/IntegerFieldValidationBinderIT.java

```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Test
+    public void required_triggerInputBlur_assertValidity() {
+        testField.sendKeys(Keys.TAB);
+        assertServerInvalid();
+        assertClientInvalid();
+        assertErrorMessage(REQUIRED_ERROR_MESSAGE);
+    }
+
+    @Test
+    public void required_changeInputValue_assertValidity() {
+        $("input").id(EXPECTED_VALUE_INPUT).sendKeys("1234", Keys.ENTER);
+
+        testField.setValue("1234");
+        assertServerValid();
+        assertClientValid();
+
+        testField.setValue("");
+        assertServerInvalid();
+        assertClientInvalid();
+        assertErrorMessage(REQUIRED_ERROR_MESSAGE);
+    }
+
+    @Test
+    public void min_changeInputValue_assertValidity() {
+        $("input").id(MIN_INPUT).sendKeys("2", Keys.ENTER);
+        $("input").id(EXPECTED_VALUE_INPUT).sendKeys("3", Keys.ENTER);
+
+        // Constraint validation fails:
+        testField.setValue("1");
+        assertClientInvalid();
+        assertServerInvalid();
+        assertErrorMessage("");
+
+        // Binder validation fails:
+        testField.setValue("2");
+        assertClientInvalid();
+        assertServerInvalid();
+        assertErrorMessage(UNEXPECTED_VALUE_ERROR_MESSAGE);
+
+        // Both validations pass:
+        testField.setValue("3");
+        assertClientValid();
+        assertServerValid();
+    }
+
+    @Test
+    public void max_changeInputValue_assertValidity() {
+        $("input").id(MAX_INPUT).sendKeys("2", Keys.ENTER);
+        $("input").id(EXPECTED_VALUE_INPUT).sendKeys("1", Keys.ENTER);
+
+        // Constraint validation fails:
+        testField.setValue("3");
+        assertClientInvalid();
+        assertServerInvalid();
+        assertErrorMessage("");
+
+        // Binder validation fails:
+        testField.setValue("2");
+        assertClientInvalid();
+        assertServerInvalid();
+        assertErrorMessage(UNEXPECTED_VALUE_ERROR_MESSAGE);
+
+        // Both validations pass:
+        testField.setValue("1");
+        assertClientValid();
+        assertServerValid();
+    }
+
+    @Test
+    public void step_changeInputValue_assertValidity() {
+        $("input").id(STEP_INPUT).sendKeys("2", Keys.ENTER);
+        $("input").id(EXPECTED_VALUE_INPUT).sendKeys("4", Keys.ENTER);
+
+        // Constraint validation fails:
+        testField.setValue("1");
+        assertClientInvalid();
+        assertServerInvalid();
+        assertErrorMessage("");
+
+        // Binder validation fails:
+        testField.setValue("2");
+        assertClientInvalid();
+        assertServerInvalid();
+        assertErrorMessage(UNEXPECTED_VALUE_ERROR_MESSAGE);
+
+        // Both validations pass:
+        testField.setValue("4");
+        assertClientValid();
+        assertServerValid();
+    }
+
+    @Test
+    public void badInput_changeInputValue_assertValidity() {
+        $("input").id(EXPECTED_VALUE_INPUT).sendKeys("2", Keys.ENTER);
+
+        testField.sendKeys("--2", Keys.TAB);
```

Comment:
**note:** Unlike other tests, I had to use <kbd>Tab</kbd> here to commit the value. <kbd>Enter</kbd> didn't work due to the issue:

- https://github.com/vaadin/web-components/issues/5113